content/blog/2025-03-20-new-hash-table.md

```diff
@@ -0,0 +1,202 @@
+  between the iterations. This is important to keep supporting the
+  [SCAN](/commands/scan/) command.
+
+* Random element sampling, for commands like [RANDOMKEY](/commands/randomkey/).
+
+These are not standard features, so we could not pick an off-the-shelf hash
+table. We had to design one ourselves.
+
+The hash table used until Valkey 8.0, called "dict", has the following memory
+layout:
+
+```
++---------+
+| dict    |          table
++---------+         +-----+-----+-----+-----+-----+-----+-----
+| table 0 --------->|  x  |  x  |  x  |  x  |  x  |  x  | ...
+| table 1 |         +-----+-----+--|--+-----+-----+-----+-----
++---------+                        |
+                                   v
+                             +-----------+         +-------+
+                             | dictEntry |    .--->| "FOO" |
+                             +-----------+   /     +-------+
+                             |    key  -----'
+                             |           |         +-------------------+
+                             |   value ----------->| serverObject      |
+                             |           |         +-------------------+
+                             |   next    |         | type, encoding,   |
+                             +-----|-----+         | ref-counter, etc. |
+                                   |               | "BAR" (embedded)  |
+                                   v               +-------------------+
+                             +-----------+
+                             | dictEntry |
+                             +-----------+
+                             |    key    |
+                             |   value   |
+                             |   next    |
+                             +-----|-----+
+                                   |
+                                   v
+                                  ...
+```
+
+The dict has two tables, called "table 0" and "table 1". Usually only one
+exists, but both are used when incremental rehashing is in progress.
+
+It is a chained hash table, so if multiple keys are hashed to the same slot in
+the table, their key-value entries form a linked list. That's what the "next"
+pointer in the dictEntry is for.
+
+To lookup a key "FOO" and access the value "BAR", Valkey still had to read from
+memory four times. If there is a hash collission, it has to follow two more
+pointers for each hash collission and thus read twice more from memory (the key
+and the next pointer).
+
+In Valkey 8.0, an optimization was made to embed the key ("FOO" in the drawing
+above) in the dictEntry, eliminating one pointer and one memory access.
+
+Design
+------
+
+In the new hash table designed for Valkey 8.1, the table consists of buckets of
+64 bytes, one cache line. Each bucket can store up to seven elements. Keys that
+map to the same bucket are all stored in the same bucket. The bucket also has a
+metadata section which contains a one byte secondary hash for each key. This is
```

Comment:
Need to explain a bit about what the secondary hash is doing here, it's not really explained what it is or how it solve hash collisions.

I'm not sure we did any testing, but this is also really good for caching workloads, since we save a memory access most of the time when there is a cache miss. A common cache hit rate is like 80%, so that 20% gets a nice speed boost as well.